Test failures with urllib3 2.0.4

#### Problem description

We are building smart-open 0.4.0 for nixpkgs, thereby running the test suite. After updating urllib3 from 1.26.16 to 2.0.4 we started seeing failures in the smart-open testsuite.

```
=========================== short test summary info ============================
FAILED smart_open/tests/test_http.py::HttpTest::test_https_seek_forward - requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRe...
FAILED smart_open/tests/test_http.py::HttpTest::test_seek_from_current - requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRe...
FAILED smart_open/tests/test_http.py::HttpTest::test_seek_from_end - requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRe...
FAILED smart_open/tests/test_http.py::HttpTest::test_seek_from_start - requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRe...
======= 4 failed, 368 passed, 4 skipped, 27 warnings in 70.45s (0:01:10) =======
```

<details><summary>Test protocol</summary>
<pre>
smart-open> _______________________ HttpTest.test_https_seek_forward _______________________
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff66e0a00>
smart-open> 
smart-open>     @contextmanager
smart-open>     def _error_catcher(self) -> typing.Generator[None, None, None]:
smart-open>         """
smart-open>         Catch low-level python exceptions, instead re-raising urllib3
smart-open>         variants, so that low-level exceptions are not leaked in the
smart-open>         high-level api.
smart-open>     
smart-open>         On exit, release the connection back to the pool.
smart-open>         """
smart-open>         clean_exit = False
smart-open>     
smart-open>         try:
smart-open>             try:
smart-open> >               yield
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:710: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff66e0a00>, amt = 131072
smart-open> 
smart-open>     def _raw_read(
smart-open>         self,
smart-open>         amt: int | None = None,
smart-open>     ) -> bytes:
smart-open>         """
smart-open>         Reads `amt` of bytes from the socket.
smart-open>         """
smart-open>         if self._fp is None:
smart-open>             return None  # type: ignore[return-value]
smart-open>     
smart-open>         fp_closed = getattr(self._fp, "closed", False)
smart-open>     
smart-open>         with self._error_catcher():
smart-open>             data = self._fp_read(amt) if not fp_closed else b""
smart-open>             if amt is not None and amt != 0 and not data:
smart-open>                 # Platform-specific: Buggy versions of Python.
smart-open>                 # Close the connection when no data is returned
smart-open>                 #
smart-open>                 # This is redundant to what httplib/http.client _should_
smart-open>                 # already do.  However, versions of python released before
smart-open>                 # December 15, 2012 (http://bugs.python.org/issue16298) do
smart-open>                 # not properly close the connection in all cases. There is
smart-open>                 # no harm in redundantly calling close.
smart-open>                 self._fp.close()
smart-open>                 if (
smart-open>                     self.enforce_content_length
smart-open>                     and self.length_remaining is not None
smart-open>                     and self.length_remaining != 0
smart-open>                 ):
smart-open>                     # This is an edge case that httplib failed to cover due
smart-open>                     # to concerns of backward compatibility. We're
smart-open>                     # addressing it here to make sure IncompleteRead is
smart-open>                     # raised during streaming, so all calls with incorrect
smart-open>                     # Content-Length are caught.
smart-open> >                   raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
smart-open> E                   urllib3.exceptions.IncompleteRead: IncompleteRead(58 bytes read, 10 more expected)
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:835: IncompleteRead
smart-open> 
smart-open> The above exception was the direct cause of the following exception:
smart-open> 
smart-open>     def generate():
smart-open>         # Special case for urllib3.
smart-open>         if hasattr(self.raw, "stream"):
smart-open>             try:
smart-open> >               yield from self.raw.stream(chunk_size, decode_content=True)
smart-open> 
smart-open> /nix/store/0ywyi98av0590s5idss3c13zas1cd1a0-python3.11-requests-2.31.0/lib/python3.11/site-packages/requests/models.py:816: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:940: in stream
smart-open>     data = self.read(amt=amt, decode_content=decode_content)
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:911: in read
smart-open>     data = self._raw_read(amt)
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:813: in _raw_read
smart-open>     with self._error_catcher():
smart-open> /nix/store/15djm2sxy19yyvicydzwrzfchq7nkmh1-python3-3.11.5/lib/python3.11/contextlib.py:155: in __exit__
smart-open>     self.gen.throw(typ, value, traceback)
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff66e0a00>
smart-open> 
smart-open>     @contextmanager
smart-open>     def _error_catcher(self) -> typing.Generator[None, None, None]:
smart-open>         """
smart-open>         Catch low-level python exceptions, instead re-raising urllib3
smart-open>         variants, so that low-level exceptions are not leaked in the
smart-open>         high-level api.
smart-open>     
smart-open>         On exit, release the connection back to the pool.
smart-open>         """
smart-open>         clean_exit = False
smart-open>     
smart-open>         try:
smart-open>             try:
smart-open>                 yield
smart-open>     
smart-open>             except SocketTimeout as e:
smart-open>                 # FIXME: Ideally we'd like to include the url in the ReadTimeoutError but
smart-open>                 # there is yet no clean way to get at it from this context.
smart-open>                 raise ReadTimeoutError(self._pool, None, "Read timed out.") from e  # type: ignore[arg-type]
smart-open>     
smart-open>             except BaseSSLError as e:
smart-open>                 # FIXME: Is there a better way to differentiate between SSLErrors?
smart-open>                 if "read operation timed out" not in str(e):
smart-open>                     # SSL errors related to framing/MAC get wrapped and reraised here
smart-open>                     raise SSLError(e) from e
smart-open>     
smart-open>                 raise ReadTimeoutError(self._pool, None, "Read timed out.") from e  # type: ignore[arg-type]
smart-open>     
smart-open>             except (HTTPException, OSError) as e:
smart-open>                 # This includes IncompleteRead.
smart-open> >               raise ProtocolError(f"Connection broken: {e!r}", e) from e
smart-open> E               urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(58 bytes read, 10 more expected)', IncompleteRead(58 bytes read, 10 more expected))
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:727: ProtocolError
smart-open> 
smart-open> During handling of the above exception, another exception occurred:
smart-open> 
smart-open> self = <smart_open.tests.test_http.HttpTest testMethod=test_https_seek_forward>
smart-open> 
smart-open>     @responses.activate
smart-open>     def test_https_seek_forward(self):
smart-open>         """Did the seek forward over HTTPS work?"""
smart-open>         responses.add_callback(responses.GET, HTTPS_URL, callback=request_callback)
smart-open>     
smart-open>         with smart_open.open(HTTPS_URL, "rb") as fin:
smart-open>             fin.seek(10)
smart-open> >           read_bytes = fin.read(size=10)
smart-open> 
smart-open> smart_open/tests/test_http.py:142: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> smart_open/http.py:183: in read
smart-open>     bytes_read = self._read_buffer.fill(self._read_iter)
smart-open> smart_open/bytebuffer.py:155: in fill
smart-open>     for more_bytes in source:
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open>     def generate():
smart-open>         # Special case for urllib3.
smart-open>         if hasattr(self.raw, "stream"):
smart-open>             try:
smart-open>                 yield from self.raw.stream(chunk_size, decode_content=True)
smart-open>             except ProtocolError as e:
smart-open> >               raise ChunkedEncodingError(e)
smart-open> E               requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(58 bytes read, 10 more expected)', IncompleteRead(58 bytes read, 10 more expected))
smart-open> 
smart-open> /nix/store/0ywyi98av0590s5idss3c13zas1cd1a0-python3.11-requests-2.31.0/lib/python3.11/site-packages/requests/models.py:818: ChunkedEncodingError
smart-open> _______________________ HttpTest.test_seek_from_current ________________________
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff2cd5570>
smart-open> 
smart-open>     @contextmanager
smart-open>     def _error_catcher(self) -> typing.Generator[None, None, None]:
smart-open>         """
smart-open>         Catch low-level python exceptions, instead re-raising urllib3
smart-open>         variants, so that low-level exceptions are not leaked in the
smart-open>         high-level api.
smart-open>     
smart-open>         On exit, release the connection back to the pool.
smart-open>         """
smart-open>         clean_exit = False
smart-open>     
smart-open>         try:
smart-open>             try:
smart-open> >               yield
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:710: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff2cd5570>, amt = 131072
smart-open> 
smart-open>     def _raw_read(
smart-open>         self,
smart-open>         amt: int | None = None,
smart-open>     ) -> bytes:
smart-open>         """
smart-open>         Reads `amt` of bytes from the socket.
smart-open>         """
smart-open>         if self._fp is None:
smart-open>             return None  # type: ignore[return-value]
smart-open>     
smart-open>         fp_closed = getattr(self._fp, "closed", False)
smart-open>     
smart-open>         with self._error_catcher():
smart-open>             data = self._fp_read(amt) if not fp_closed else b""
smart-open>             if amt is not None and amt != 0 and not data:
smart-open>                 # Platform-specific: Buggy versions of Python.
smart-open>                 # Close the connection when no data is returned
smart-open>                 #
smart-open>                 # This is redundant to what httplib/http.client _should_
smart-open>                 # already do.  However, versions of python released before
smart-open>                 # December 15, 2012 (http://bugs.python.org/issue16298) do
smart-open>                 # not properly close the connection in all cases. There is
smart-open>                 # no harm in redundantly calling close.
smart-open>                 self._fp.close()
smart-open>                 if (
smart-open>                     self.enforce_content_length
smart-open>                     and self.length_remaining is not None
smart-open>                     and self.length_remaining != 0
smart-open>                 ):
smart-open>                     # This is an edge case that httplib failed to cover due
smart-open>                     # to concerns of backward compatibility. We're
smart-open>                     # addressing it here to make sure IncompleteRead is
smart-open>                     # raised during streaming, so all calls with incorrect
smart-open>                     # Content-Length are caught.
smart-open> >                   raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
smart-open> E                   urllib3.exceptions.IncompleteRead: IncompleteRead(58 bytes read, 10 more expected)
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:835: IncompleteRead
smart-open> 
smart-open> The above exception was the direct cause of the following exception:
smart-open> 
smart-open>     def generate():
smart-open>         # Special case for urllib3.
smart-open>         if hasattr(self.raw, "stream"):
smart-open>             try:
smart-open> >               yield from self.raw.stream(chunk_size, decode_content=True)
smart-open> 
smart-open> /nix/store/0ywyi98av0590s5idss3c13zas1cd1a0-python3.11-requests-2.31.0/lib/python3.11/site-packages/requests/models.py:816: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:940: in stream
smart-open>     data = self.read(amt=amt, decode_content=decode_content)
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:911: in read
smart-open>     data = self._raw_read(amt)
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:813: in _raw_read
smart-open>     with self._error_catcher():
smart-open> /nix/store/15djm2sxy19yyvicydzwrzfchq7nkmh1-python3-3.11.5/lib/python3.11/contextlib.py:155: in __exit__
smart-open>     self.gen.throw(typ, value, traceback)
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff2cd5570>
smart-open> 
smart-open>     @contextmanager
smart-open>     def _error_catcher(self) -> typing.Generator[None, None, None]:
smart-open>         """
smart-open>         Catch low-level python exceptions, instead re-raising urllib3
smart-open>         variants, so that low-level exceptions are not leaked in the
smart-open>         high-level api.
smart-open>     
smart-open>         On exit, release the connection back to the pool.
smart-open>         """
smart-open>         clean_exit = False
smart-open>     
smart-open>         try:
smart-open>             try:
smart-open>                 yield
smart-open>     
smart-open>             except SocketTimeout as e:
smart-open>                 # FIXME: Ideally we'd like to include the url in the ReadTimeoutError but
smart-open>                 # there is yet no clean way to get at it from this context.
smart-open>                 raise ReadTimeoutError(self._pool, None, "Read timed out.") from e  # type: ignore[arg-type]
smart-open>     
smart-open>             except BaseSSLError as e:
smart-open>                 # FIXME: Is there a better way to differentiate between SSLErrors?
smart-open>                 if "read operation timed out" not in str(e):
smart-open>                     # SSL errors related to framing/MAC get wrapped and reraised here
smart-open>                     raise SSLError(e) from e
smart-open>     
smart-open>                 raise ReadTimeoutError(self._pool, None, "Read timed out.") from e  # type: ignore[arg-type]
smart-open>     
smart-open>             except (HTTPException, OSError) as e:
smart-open>                 # This includes IncompleteRead.
smart-open> >               raise ProtocolError(f"Connection broken: {e!r}", e) from e
smart-open> E               urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(58 bytes read, 10 more expected)', IncompleteRead(58 bytes read, 10 more expected))
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:727: ProtocolError
smart-open> 
smart-open> During handling of the above exception, another exception occurred:
smart-open> 
smart-open> self = <smart_open.tests.test_http.HttpTest testMethod=test_seek_from_current>
smart-open> 
smart-open>     @responses.activate
smart-open>     def test_seek_from_current(self):
smart-open>         responses.add_callback(responses.GET, URL, callback=request_callback)
smart-open>         reader = smart_open.http.SeekableBufferedInputBase(URL)
smart-open>     
smart-open>         reader.seek(10)
smart-open> >       read_bytes = reader.read(size=10)
smart-open> 
smart-open> smart_open/tests/test_http.py:79: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> smart_open/http.py:183: in read
smart-open>     bytes_read = self._read_buffer.fill(self._read_iter)
smart-open> smart_open/bytebuffer.py:155: in fill
smart-open>     for more_bytes in source:
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open>     def generate():
smart-open>         # Special case for urllib3.
smart-open>         if hasattr(self.raw, "stream"):
smart-open>             try:
smart-open>                 yield from self.raw.stream(chunk_size, decode_content=True)
smart-open>             except ProtocolError as e:
smart-open> >               raise ChunkedEncodingError(e)
smart-open> E               requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(58 bytes read, 10 more expected)', IncompleteRead(58 bytes read, 10 more expected))
smart-open> 
smart-open> /nix/store/0ywyi98av0590s5idss3c13zas1cd1a0-python3.11-requests-2.31.0/lib/python3.11/site-packages/requests/models.py:818: ChunkedEncodingError
smart-open> _________________________ HttpTest.test_seek_from_end __________________________
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff289d210>
smart-open> 
smart-open>     @contextmanager
smart-open>     def _error_catcher(self) -> typing.Generator[None, None, None]:
smart-open>         """
smart-open>         Catch low-level python exceptions, instead re-raising urllib3
smart-open>         variants, so that low-level exceptions are not leaked in the
smart-open>         high-level api.
smart-open>     
smart-open>         On exit, release the connection back to the pool.
smart-open>         """
smart-open>         clean_exit = False
smart-open>     
smart-open>         try:
smart-open>             try:
smart-open> >               yield
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:710: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff289d210>, amt = 131072
smart-open> 
smart-open>     def _raw_read(
smart-open>         self,
smart-open>         amt: int | None = None,
smart-open>     ) -> bytes:
smart-open>         """
smart-open>         Reads `amt` of bytes from the socket.
smart-open>         """
smart-open>         if self._fp is None:
smart-open>             return None  # type: ignore[return-value]
smart-open>     
smart-open>         fp_closed = getattr(self._fp, "closed", False)
smart-open>     
smart-open>         with self._error_catcher():
smart-open>             data = self._fp_read(amt) if not fp_closed else b""
smart-open>             if amt is not None and amt != 0 and not data:
smart-open>                 # Platform-specific: Buggy versions of Python.
smart-open>                 # Close the connection when no data is returned
smart-open>                 #
smart-open>                 # This is redundant to what httplib/http.client _should_
smart-open>                 # already do.  However, versions of python released before
smart-open>                 # December 15, 2012 (http://bugs.python.org/issue16298) do
smart-open>                 # not properly close the connection in all cases. There is
smart-open>                 # no harm in redundantly calling close.
smart-open>                 self._fp.close()
smart-open>                 if (
smart-open>                     self.enforce_content_length
smart-open>                     and self.length_remaining is not None
smart-open>                     and self.length_remaining != 0
smart-open>                 ):
smart-open>                     # This is an edge case that httplib failed to cover due
smart-open>                     # to concerns of backward compatibility. We're
smart-open>                     # addressing it here to make sure IncompleteRead is
smart-open>                     # raised during streaming, so all calls with incorrect
smart-open>                     # Content-Length are caught.
smart-open> >                   raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
smart-open> E                   urllib3.exceptions.IncompleteRead: IncompleteRead(10 bytes read, 58 more expected)
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:835: IncompleteRead
smart-open> 
smart-open> The above exception was the direct cause of the following exception:
smart-open> 
smart-open>     def generate():
smart-open>         # Special case for urllib3.
smart-open>         if hasattr(self.raw, "stream"):
smart-open>             try:
smart-open> >               yield from self.raw.stream(chunk_size, decode_content=True)
smart-open> 
smart-open> /nix/store/0ywyi98av0590s5idss3c13zas1cd1a0-python3.11-requests-2.31.0/lib/python3.11/site-packages/requests/models.py:816: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:940: in stream
smart-open>     data = self.read(amt=amt, decode_content=decode_content)
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:911: in read
smart-open>     data = self._raw_read(amt)
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:813: in _raw_read
smart-open>     with self._error_catcher():
smart-open> /nix/store/15djm2sxy19yyvicydzwrzfchq7nkmh1-python3-3.11.5/lib/python3.11/contextlib.py:155: in __exit__
smart-open>     self.gen.throw(typ, value, traceback)
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff289d210>
smart-open> 
smart-open>     @contextmanager
smart-open>     def _error_catcher(self) -> typing.Generator[None, None, None]:
smart-open>         """
smart-open>         Catch low-level python exceptions, instead re-raising urllib3
smart-open>         variants, so that low-level exceptions are not leaked in the
smart-open>         high-level api.
smart-open>     
smart-open>         On exit, release the connection back to the pool.
smart-open>         """
smart-open>         clean_exit = False
smart-open>     
smart-open>         try:
smart-open>             try:
smart-open>                 yield
smart-open>     
smart-open>             except SocketTimeout as e:
smart-open>                 # FIXME: Ideally we'd like to include the url in the ReadTimeoutError but
smart-open>                 # there is yet no clean way to get at it from this context.
smart-open>                 raise ReadTimeoutError(self._pool, None, "Read timed out.") from e  # type: ignore[arg-type]
smart-open>     
smart-open>             except BaseSSLError as e:
smart-open>                 # FIXME: Is there a better way to differentiate between SSLErrors?
smart-open>                 if "read operation timed out" not in str(e):
smart-open>                     # SSL errors related to framing/MAC get wrapped and reraised here
smart-open>                     raise SSLError(e) from e
smart-open>     
smart-open>                 raise ReadTimeoutError(self._pool, None, "Read timed out.") from e  # type: ignore[arg-type]
smart-open>     
smart-open>             except (HTTPException, OSError) as e:
smart-open>                 # This includes IncompleteRead.
smart-open> >               raise ProtocolError(f"Connection broken: {e!r}", e) from e
smart-open> E               urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(10 bytes read, 58 more expected)', IncompleteRead(10 bytes read, 58 more expected))
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:727: ProtocolError
smart-open> 
smart-open> During handling of the above exception, another exception occurred:
smart-open> 
smart-open> self = <smart_open.tests.test_http.HttpTest testMethod=test_seek_from_end>
smart-open> 
smart-open>     @responses.activate
smart-open>     def test_seek_from_end(self):
smart-open>         responses.add_callback(responses.GET, URL, callback=request_callback)
smart-open>         reader = smart_open.http.SeekableBufferedInputBase(URL)
smart-open>     
smart-open>         reader.seek(-10, whence=smart_open.constants.WHENCE_END)
smart-open>         self.assertEqual(reader.tell(), len(BYTES) - 10)
smart-open> >       read_bytes = reader.read(size=10)
smart-open> 
smart-open> smart_open/tests/test_http.py:96: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> smart_open/http.py:183: in read
smart-open>     bytes_read = self._read_buffer.fill(self._read_iter)
smart-open> smart_open/bytebuffer.py:155: in fill
smart-open>     for more_bytes in source:
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open>     def generate():
smart-open>         # Special case for urllib3.
smart-open>         if hasattr(self.raw, "stream"):
smart-open>             try:
smart-open>                 yield from self.raw.stream(chunk_size, decode_content=True)
smart-open>             except ProtocolError as e:
smart-open> >               raise ChunkedEncodingError(e)
smart-open> E               requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(10 bytes read, 58 more expected)', IncompleteRead(10 bytes read, 58 more expected))
smart-open> 
smart-open> /nix/store/0ywyi98av0590s5idss3c13zas1cd1a0-python3.11-requests-2.31.0/lib/python3.11/site-packages/requests/models.py:818: ChunkedEncodingError
smart-open> ________________________ HttpTest.test_seek_from_start _________________________
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff332b340>
smart-open> 
smart-open>     @contextmanager
smart-open>     def _error_catcher(self) -> typing.Generator[None, None, None]:
smart-open>         """
smart-open>         Catch low-level python exceptions, instead re-raising urllib3
smart-open>         variants, so that low-level exceptions are not leaked in the
smart-open>         high-level api.
smart-open>     
smart-open>         On exit, release the connection back to the pool.
smart-open>         """
smart-open>         clean_exit = False
smart-open>     
smart-open>         try:
smart-open>             try:
smart-open> >               yield
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:710: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff332b340>, amt = 131072
smart-open> 
smart-open>     def _raw_read(
smart-open>         self,
smart-open>         amt: int | None = None,
smart-open>     ) -> bytes:
smart-open>         """
smart-open>         Reads `amt` of bytes from the socket.
smart-open>         """
smart-open>         if self._fp is None:
smart-open>             return None  # type: ignore[return-value]
smart-open>     
smart-open>         fp_closed = getattr(self._fp, "closed", False)
smart-open>     
smart-open>         with self._error_catcher():
smart-open>             data = self._fp_read(amt) if not fp_closed else b""
smart-open>             if amt is not None and amt != 0 and not data:
smart-open>                 # Platform-specific: Buggy versions of Python.
smart-open>                 # Close the connection when no data is returned
smart-open>                 #
smart-open>                 # This is redundant to what httplib/http.client _should_
smart-open>                 # already do.  However, versions of python released before
smart-open>                 # December 15, 2012 (http://bugs.python.org/issue16298) do
smart-open>                 # not properly close the connection in all cases. There is
smart-open>                 # no harm in redundantly calling close.
smart-open>                 self._fp.close()
smart-open>                 if (
smart-open>                     self.enforce_content_length
smart-open>                     and self.length_remaining is not None
smart-open>                     and self.length_remaining != 0
smart-open>                 ):
smart-open>                     # This is an edge case that httplib failed to cover due
smart-open>                     # to concerns of backward compatibility. We're
smart-open>                     # addressing it here to make sure IncompleteRead is
smart-open>                     # raised during streaming, so all calls with incorrect
smart-open>                     # Content-Length are caught.
smart-open> >                   raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
smart-open> E                   urllib3.exceptions.IncompleteRead: IncompleteRead(58 bytes read, 10 more expected)
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:835: IncompleteRead
smart-open> 
smart-open> The above exception was the direct cause of the following exception:
smart-open> 
smart-open>     def generate():
smart-open>         # Special case for urllib3.
smart-open>         if hasattr(self.raw, "stream"):
smart-open>             try:
smart-open> >               yield from self.raw.stream(chunk_size, decode_content=True)
smart-open> 
smart-open> /nix/store/0ywyi98av0590s5idss3c13zas1cd1a0-python3.11-requests-2.31.0/lib/python3.11/site-packages/requests/models.py:816: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:940: in stream
smart-open>     data = self.read(amt=amt, decode_content=decode_content)
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:911: in read
smart-open>     data = self._raw_read(amt)
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:813: in _raw_read
smart-open>     with self._error_catcher():
smart-open> /nix/store/15djm2sxy19yyvicydzwrzfchq7nkmh1-python3-3.11.5/lib/python3.11/contextlib.py:155: in __exit__
smart-open>     self.gen.throw(typ, value, traceback)
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open> self = <urllib3.response.HTTPResponse object at 0x7ffff332b340>
smart-open> 
smart-open>     @contextmanager
smart-open>     def _error_catcher(self) -> typing.Generator[None, None, None]:
smart-open>         """
smart-open>         Catch low-level python exceptions, instead re-raising urllib3
smart-open>         variants, so that low-level exceptions are not leaked in the
smart-open>         high-level api.
smart-open>     
smart-open>         On exit, release the connection back to the pool.
smart-open>         """
smart-open>         clean_exit = False
smart-open>     
smart-open>         try:
smart-open>             try:
smart-open>                 yield
smart-open>     
smart-open>             except SocketTimeout as e:
smart-open>                 # FIXME: Ideally we'd like to include the url in the ReadTimeoutError but
smart-open>                 # there is yet no clean way to get at it from this context.
smart-open>                 raise ReadTimeoutError(self._pool, None, "Read timed out.") from e  # type: ignore[arg-type]
smart-open>     
smart-open>             except BaseSSLError as e:
smart-open>                 # FIXME: Is there a better way to differentiate between SSLErrors?
smart-open>                 if "read operation timed out" not in str(e):
smart-open>                     # SSL errors related to framing/MAC get wrapped and reraised here
smart-open>                     raise SSLError(e) from e
smart-open>     
smart-open>                 raise ReadTimeoutError(self._pool, None, "Read timed out.") from e  # type: ignore[arg-type]
smart-open>     
smart-open>             except (HTTPException, OSError) as e:
smart-open>                 # This includes IncompleteRead.
smart-open> >               raise ProtocolError(f"Connection broken: {e!r}", e) from e
smart-open> E               urllib3.exceptions.ProtocolError: ('Connection broken: IncompleteRead(58 bytes read, 10 more expected)', IncompleteRead(58 bytes read, 10 more expected))
smart-open> 
smart-open> /nix/store/dxqvdy0rggdg8zqja3m3wf34dx5wqq3q-python3.11-urllib3-2.0.4/lib/python3.11/site-packages/urllib3/response.py:727: ProtocolError
smart-open> 
smart-open> During handling of the above exception, another exception occurred:
smart-open> 
smart-open> self = <smart_open.tests.test_http.HttpTest testMethod=test_seek_from_start>
smart-open> 
smart-open>     @responses.activate
smart-open>     def test_seek_from_start(self):
smart-open>         responses.add_callback(responses.GET, URL, callback=request_callback)
smart-open>         reader = smart_open.http.SeekableBufferedInputBase(URL)
smart-open>     
smart-open>         reader.seek(10)
smart-open>         self.assertEqual(reader.tell(), 10)
smart-open> >       read_bytes = reader.read(size=10)
smart-open> 
smart-open> smart_open/tests/test_http.py:61: 
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> smart_open/http.py:183: in read
smart-open>     bytes_read = self._read_buffer.fill(self._read_iter)
smart-open> smart_open/bytebuffer.py:155: in fill
smart-open>     for more_bytes in source:
smart-open> _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
smart-open> 
smart-open>     def generate():
smart-open>         # Special case for urllib3.
smart-open>         if hasattr(self.raw, "stream"):
smart-open>             try:
smart-open>                 yield from self.raw.stream(chunk_size, decode_content=True)
smart-open>             except ProtocolError as e:
smart-open> >               raise ChunkedEncodingError(e)
smart-open> E               requests.exceptions.ChunkedEncodingError: ('Connection broken: IncompleteRead(58 bytes read, 10 more expected)', IncompleteRead(58 bytes read, 10 more expected))
smart-open> 
smart-open> /nix/store/0ywyi98av0590s5idss3c13zas1cd1a0-python3.11-requests-2.31.0/lib/python3.11/site-packages/requests/models.py:818: ChunkedEncodingError
</pre>
</details>

#### Steps/code to reproduce the problem

Install requests, with urllib3>2 and run the test suite.

#### Versions

- Linux-6.1.46-x86_64-with-glibc2.37
- Python 3.11.5 (main, Aug 24 2023, 12:23:19) [GCC 12.3.0]
- smart_open 6.4.0

#### Checklist

Before you create the issue, please make sure you have:

- [x] Described the problem clearly
- [x] Provided a minimal reproducible example, including any required data
- [x] Provided the version numbers of the relevant software

--- a/.github/workflows/python-package.yml
+++ b/.github/workflows/python-package.yml
@@ -1,5 +1,8 @@
 name: Test
 on: [push, pull_request]
+concurrency:  # https://stackoverflow.com/questions/66335225#comment133398800_72408109
+  group: ${{ github.workflow }}-${{ github.ref || github.run_id }}
+  cancel-in-progress: ${{ github.event_name == 'pull_request' }}
 jobs:
   linters:
     runs-on: ubuntu-latest
@@ -12,10 +15,10 @@ jobs:
           python-version: "3.11"
 
       - name: Update pip
-        run: python -m pip install -U pip
+        run: pip install -U pip
 
       - name: Install dependencies
-        run: python -m pip install flake8
+        run: pip install flake8
 
       - name: Run flake8 linter (source)
         run: flake8 --show-source smart_open
@@ -26,17 +29,17 @@ jobs:
     strategy:
       matrix:
         include:
-          - {python: '3.8', os: ubuntu-20.04}
-          - {python: '3.9', os: ubuntu-20.04}
-          - {python: '3.10', os: ubuntu-20.04}
-          - {python: '3.11', os: ubuntu-20.04}
-          - {python: '3.12', os: ubuntu-20.04}
+          - {python-version: '3.8', os: ubuntu-20.04}
+          - {python-version: '3.9', os: ubuntu-20.04}
+          - {python-version: '3.10', os: ubuntu-20.04}
+          - {python-version: '3.11', os: ubuntu-20.04}
+          - {python-version: '3.12', os: ubuntu-20.04}
 
-          - {python: '3.8', os: windows-2019}
-          - {python: '3.9', os: windows-2019}
-          - {python: '3.10', os: windows-2019}
-          - {python: '3.11', os: windows-2019}
-          - {python: '3.12', os: windows-2019}
+          - {python-version: '3.8', os: windows-2019}
+          - {python-version: '3.9', os: windows-2019}
+          - {python-version: '3.10', os: windows-2019}
+          - {python-version: '3.11', os: windows-2019}
+          - {python-version: '3.12', os: windows-2019}
     steps:
       - uses: actions/checkout@v2
 
@@ -45,13 +48,7 @@ jobs:
           python-version: ${{ matrix.python-version }}
 
       - name: Update pip
-        run: python -m pip install -U pip
-
-      #
-      # https://askubuntu.com/questions/1428181/module-lib-has-no-attribute-x509-v-flag-cb-issuer-check
-      #
-      - name: Upgrade PyOpenSSL
-        run: python -m pip install pyOpenSSL --upgrade
+        run: pip install -U pip
 
       - name: Install smart_open without dependencies
         run: pip install -e .
@@ -71,20 +68,20 @@ jobs:
     strategy:
       matrix:
         include:
-          - {python: '3.8', os: ubuntu-20.04}
-          - {python: '3.9', os: ubuntu-20.04}
-          - {python: '3.10', os: ubuntu-20.04}
-          - {python: '3.11', os: ubuntu-20.04}
-          - {python: '3.12', os: ubuntu-20.04}
+          - {python-version: '3.8', os: ubuntu-20.04}
+          - {python-version: '3.9', os: ubuntu-20.04}
+          - {python-version: '3.10', os: ubuntu-20.04}
+          - {python-version: '3.11', os: ubuntu-20.04}
+          - {python-version: '3.12', os: ubuntu-20.04}
 
           #
           # Some of the doctests don't pass on Windows because of Windows-specific
           # character encoding issues.
           #
-          # - {python: '3.7', os: windows-2019}
-          # - {python: '3.8', os: windows-2019}
-          # - {python: '3.9', os: windows-2019}
-          # - {python: '3.10', os: windows-2019}
+          # - {python-version: '3.7', os: windows-2019}
+          # - {python-version: '3.8', os: windows-2019}
+          # - {python-version: '3.9', os: windows-2019}
+          # - {python-version: '3.10', os: windows-2019}
 
     steps:
       - uses: actions/checkout@v2
@@ -94,10 +91,7 @@ jobs:
           python-version: ${{ matrix.python-version }}
 
       - name: Update pip
-        run: python -m pip install -U pip
-
-      - name: Upgrade PyOpenSSL
-        run: python -m pip install pyOpenSSL --upgrade
+        run: pip install -U pip
 
       - name: Install smart_open and its dependencies
         run: pip install -e .[test]
@@ -114,18 +108,18 @@ jobs:
     strategy:
       matrix:
         include:
-          - {python: '3.8', os: ubuntu-20.04}
-          - {python: '3.9', os: ubuntu-20.04}
-          - {python: '3.10', os: ubuntu-20.04}
-          - {python: '3.11', os: ubuntu-20.04}
-          - {python: '3.12', os: ubuntu-20.04}
+          - {python-version: '3.8', os: ubuntu-20.04}
+          - {python-version: '3.9', os: ubuntu-20.04}
+          - {python-version: '3.10', os: ubuntu-20.04}
+          - {python-version: '3.11', os: ubuntu-20.04}
+          - {python-version: '3.12', os: ubuntu-20.04}
 
           # Not sure why we exclude these, perhaps for historical reasons?
           #
-          # - {python: '3.7', os: windows-2019}
-          # - {python: '3.8', os: windows-2019}
-          # - {python: '3.9', os: windows-2019}
-          # - {python: '3.10', os: windows-2019}
+          # - {python-version: '3.7', os: windows-2019}
+          # - {python-version: '3.8', os: windows-2019}
+          # - {python-version: '3.9', os: windows-2019}
+          # - {python-version: '3.10', os: windows-2019}
 
     steps:
       - uses: actions/checkout@v2
@@ -135,12 +129,7 @@ jobs:
           python-version: ${{ matrix.python-version }}
 
       - name: Update pip
-        run: python -m pip install -U pip
-
-      - name: Upgrade PyOpenSSL
-        run: python -m pip install pyOpenSSL --upgrade
-
-      - run: python -m pip install numpy
+        run: pip install -U pip
 
       - name: Install smart_open and its dependencies
         run: pip install -e .[test]
@@ -148,7 +137,9 @@ jobs:
       - run: bash ci_helpers/helpers.sh enable_moto_server
         if: ${{ matrix.moto_server }}
 
-      - run: |
+      - name: Start vsftpd
+        timeout-minutes: 2
+        run: |
           sudo apt-get install vsftpd
           sudo bash ci_helpers/helpers.sh create_ftp_ftps_servers
 
@@ -169,16 +160,16 @@ jobs:
     strategy:
       matrix:
         include:
-          - {python: '3.8', os: ubuntu-20.04}
-          - {python: '3.9', os: ubuntu-20.04}
-          - {python: '3.10', os: ubuntu-20.04}
-          - {python: '3.11', os: ubuntu-20.04}
-          - {python: '3.12', os: ubuntu-20.04}
+          - {python-version: '3.8', os: ubuntu-20.04}
+          - {python-version: '3.9', os: ubuntu-20.04}
+          - {python-version: '3.10', os: ubuntu-20.04}
+          - {python-version: '3.11', os: ubuntu-20.04}
+          - {python-version: '3.12', os: ubuntu-20.04}
 
-          # - {python: '3.7', os: windows-2019}
-          # - {python: '3.8', os: windows-2019}
-          # - {python: '3.9', os: windows-2019}
-          # - {python: '3.10', os: windows-2019}
+          # - {python-version: '3.7', os: windows-2019}
+          # - {python-version: '3.8', os: windows-2019}
+          # - {python-version: '3.9', os: windows-2019}
+          # - {python-version: '3.10', os: windows-2019}
 
     steps:
       - uses: actions/checkout@v2
@@ -188,15 +179,10 @@ jobs:
           python-version: ${{ matrix.python-version }}
 
       - name: Update pip
-        run: python -m pip install -U pip
-
-      - name: Upgrade PyOpenSSL
-        run: python -m pip install pyOpenSSL --upgrade
+        run: pip install -U pip
 
       - name: Install smart_open and its dependencies
         run: pip install -e .[test]
-
-      - run: pip install awscli pytest_benchmark
 
       - name: Run benchmarks
         run: python ci_helpers/run_benchmarks.py

--- a/ci_helpers/helpers.sh
+++ b/ci_helpers/helpers.sh
@@ -20,6 +20,9 @@ create_ftp_ftps_servers(){
   mkdir $home_dir
   useradd -p $(echo $pass | openssl passwd -1 -stdin) -d $home_dir $user
   chown $user:$user $home_dir
+  openssl req -x509 -nodes -new -sha256 -days 10240 -newkey rsa:2048 -keyout /etc/vsftpd.key -out /etc/vsftpd.pem -subj "/C=ZA/CN=localhost"
+  chmod 755 /etc/vsftpd.key
+  chmod 755 /etc/vsftpd.pem
 
   server_setup='''
 listen=YES
@@ -32,6 +35,8 @@ chroot_local_user=YES
 allow_writeable_chroot=YES'''
 
   additional_ssl_setup='''
+rsa_cert_file=/etc/vsftpd.pem
+rsa_private_key_file=/etc/vsftpd.key
 ssl_enable=YES
 allow_anon_ssl=NO
 force_local_data_ssl=NO

--- a/integration-tests/test_ftp.py
+++ b/integration-tests/test_ftp.py
@@ -1,6 +1,11 @@
 from __future__ import unicode_literals
 import pytest
 from smart_open import open
+import ssl
+from functools import partial
+
+# localhost has self-signed cert, see ci_helpers/helpers.sh:create_ftp_ftps_servers
+ssl.create_default_context = partial(ssl.create_default_context, cafile="/etc/vsftpd.pem")
 
 
 @pytest.fixture(params=[("ftp", 21), ("ftps", 90)])

--- a/setup.py
+++ b/setup.py
@@ -47,9 +47,12 @@ all_deps = aws_deps + gcs_deps + azure_deps + http_deps + ssh_deps + zst_deps
 tests_require = all_deps + [
     'moto[server]',
     'responses',
-    'boto3',
     'pytest',
     'pytest-rerunfailures',
+    'pytest_benchmark',
+    'awscli',
+    'pyopenssl',
+    'numpy',
 ]
 
 setup(

--- a/smart_open/ftp.py
+++ b/smart_open/ftp.py
@@ -86,7 +86,7 @@ def convert_transport_params_to_args(transport_params):
 def _connect(hostname, username, port, password, secure_connection, transport_params):
     kwargs = convert_transport_params_to_args(transport_params)
     if secure_connection:
-        ssl_context = ssl.create_default_context(purpose=ssl.Purpose.CLIENT_AUTH)
+        ssl_context = ssl.create_default_context(purpose=ssl.Purpose.SERVER_AUTH)
         ftp = FTP_TLS(context=ssl_context, **kwargs)
     else:
         ftp = FTP(**kwargs)

--- a/smart_open/tests/test_http.py
+++ b/smart_open/tests/test_http.py
@@ -21,24 +21,22 @@ BYTES = b'i tried so hard and got so far but in the end it doesn\'t even matter'
 URL = 'http://localhost'
 HTTPS_URL = 'https://localhost'
 HEADERS = {
-    'Content-Length': str(len(BYTES)),
     'Accept-Ranges': 'bytes',
 }
 
 
 def request_callback(request, headers=HEADERS, data=BYTES):
-    try:
-        range_string = request.headers['range']
-    except KeyError:
-        return (200, headers, data)
+    headers = headers.copy()
+    range_string = request.headers.get('range', 'bytes=0-')
 
-    start, end = range_string.replace('bytes=', '').split('-', 1)
+    start, end = range_string.replace('bytes=', '', 1).split('-', 1)
     start = int(start)
-    if end:
-        end = int(end)
-    else:
-        end = len(data)
-    return (200, headers, data[start:end])
+    end = int(end) if end else len(data)
+
+    data = data[start:end]
+    headers['Content-Length'] = str(len(data))
+
+    return (200, headers, data)
 
 
 @unittest.skipIf(os.environ.get('TRAVIS'), 'This test does not work on TravisCI for some reason')

